app/models/version.rb

```diff
@@ -393,4 +394,9 @@ def metadata_attribute_length
       errors.add(:metadata, "metadata key is empty") if key.empty?
     end
   end
+
+  def unique_canonical_number
+    version = Version.find_by(canonical_number: canonical_number, rubygem_id: rubygem_id, platform: platform)
```

Comment:
If I understand it well, this can (when not using `find_by`) return multiple results now. Wouldn't be better to show all versions then?

Reply:
It could, for the current 179 duplications that we have. I think it's very unlikely that any of them will try to publish yet another duplication of the same version, but I guess showing all versions does hurt anyways?

Reply:
I'm just trying to find the balance between things we can rely on (data normalization) and the real state. We don't enforce this uniqueness on DB level so it can still happen since DB is our source of truth.

:thinking: what about to introduce some flag to (true/false) and enforce uniqueness this only for new gems (where true would become default)? That way we can postpone decision what to do with those wrong versions, but already enforce this on new ones.

Reply:
Isn't that exactly what this PR is doing? :thinking:

If the DB doesn't meet the uniqueness requirement, how can we set a uniqueness DB constraint? With this PR we are enforcing uniqueness via model validations upon creation, which I think is what you're asking for?

Reply:
I had an idea.

How about setting a wrong (but unique) canonical version for the affected duplications (for example, the version itself)? This way we still allow publishing yet another duplication of those gems (which are messed up anyways), but enforce normalized data in exchange.

Does that make sense?

Reply:
@deivid-rodriguez it is possible to introduce DB uniqueness requirement partially and make it work for new versions by default.

It should be possible in this way (not tested, just pseudo-code):

```ruby
# make it default for new versions
add_column :versions, :strict_versioning, :boolean, default: true
# update all versions having problem with constraint to strict_versioning: false
Version.with_versioning_dupes.update_all(strict_versioning: false)
# add unique index to cover only strict_versioned versions
execute("CREATE UNIQUE INDEX enforce_strict_canonical_number ON versions USING(semantic_version) WHERE strict_versioning = true")
```

Reply:
Oh, I see what you meant now, add yet another column to signal this.

I think it's essentially the same idea as mine although mine doesn't need a new column.

Reply:
I'm happy with your suggestion! :+1:

Reply:
> How about setting a wrong (but unique) canonical version for the affected duplications (for example, the <version>.<gem_name>.<gem_platform>.ignored, or something like that)?

I like this idea. We can selectively use `ignored` so that only half of the affected versions have this value (keeping the entire set of cannonical_numbers unique).
@simi what do you think? Do you see any issue with this? The backfill task will be slightly more complicated and we will need to add the index in a follow-up.

Reply:
I do prefer to keep data clean (normalised) as possible. For `semantic_version` column I would expect all values to have the same logic. Introducing another column value logic (like `<gem_name>.<gem_platform>.ignored`) is similar for me as (for example) storing array in text column separated by semicolon.

Additional column is there just to make it possible to temporarily enforce uniqueness even with legacy data. The best would be to just decide somehow to normalize legacy data as well. But thanks to this additional column this could be done later.

But any solution ending up with unique index would fit in here.

Reply:
What I like about @simi's proposal is that it's "cleaner". However, my gut tells me that this data normalization might never actually happen, so by implementing that solution we're bound to keeping an extra column that we don't really need forever. If we come up with a good strategy to clean this data, then I think it's fine, but then... we can maybe clean it right away? If we need to delay coming up with a strategy because we can't think of any, then I think it's better to go with my alternative.

Thinking of strategies, some manual testing indicates that one of the duplicated versions seems always preferred over the other (maybe because it comes first in the index or something?). So there's really no way to install the other, it seems? If this is the case, maybe we can actually yank the versions that can't be installed?

```
$ gem install faker:2.0.0
Successfully installed faker-2.0
1 gem installed
deivid@chaba ~/Code/playground/faker $ gem install faker:2.0
Successfully installed faker-2.0
1 gem installed
```

Reply:
>  So there's really no way to install the other, it seems? If this is the case, maybe we can actually yank the versions that can't be installed?

faker 2.0.0 can be installed using Gemfile. make sure you haven't installed faker 2.0 already.

```
# update all versions having problem with constraint to strict_versioning: false
Version.with_versioning_dupes.update_all(strict_versioning: false)
# add unique index to cover only strict_versioned versions
execute("CREATE UNIQUE INDEX enforce_strict_canonical_number ON versions USING(semantic_version) WHERE strict_versioning = true")
```
note that this doesn't add unique constrain on the existing versions. if we were to add a separate column, `strict_versioning` would be false just for 179 duplicates we have (not all existing versions).
I agree that our data would not be normal for those 179 rows but I think we can treat them as exceptions and not change our schema to accommodate them (too big a price).

Reply:
There's a chance that the currently non-installable version got installed during the time window where it was actually installable (before the duplication was created), and it's hard to make sure that these gems really have the same code (and they still differ at least in their version constants). So... caring about things feels too complicated for little benefit.

Reply:
Sorry, I didn't read your reply before posting mine, second time it happens to me today :grimacing:.

I can't seem to be able to install faker 2.0.0:

```
$ cat Gemfile && gem uninstall faker --all --executables && rm -f Gemfile.lock && bundle install && gem list fake
source "https://rubygems.org"

gem "faker", "2.0.0"
Gem 'faker' is not installed
Fetching gem metadata from https://rubygems.org/....
Resolving dependencies...
Using concurrent-ruby 1.1.7
Using bundler 2.2.0.rc.2
Using i18n 1.8.5
Fetching faker 2.0
Installing faker 2.0
Bundle complete! 1 Gemfile dependency, 4 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.

*** LOCAL GEMS ***

faker (2.0)
```

In any case, my message was suggesting to go with my proposal even if the "uninstallable hypothesis" was true, so it doesn't really matter :smile:.

> note that this doesn't add unique constrain on the existing versions. if we were to add a separate column, strict_versioning would be false just for 179 duplicates we have (not all existing versions).

Yep, I think all three of us are aware of this. But we think this is too much for little benefit, whereas @simi prefers to not fake that we have unique canonical data when we don't have it.

Anwyays, since @simi said he's ok with any solution that enforces unique canonical versions in DB going forward, my approach seems ok :+1:.

Reply:
Try this:
```
sudo docker run -it ruby:2.7-alpine /bin/sh -c "printf \"source 'https://rubygems.org'\ngem 'faker', '2.0.0'\n\" > Gemfile && bundle install"
Fetching gem metadata from https://rubygems.org/....
Resolving dependencies...
Using bundler 2.1.4
Fetching concurrent-ruby 1.1.7
Installing concurrent-ruby 1.1.7
Fetching i18n 1.8.5
Installing i18n 1.8.5
Fetching faker 2.0.0
Installing faker 2.0.0
Bundle complete! 1 Gemfile dependency, 4 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
Post-install message from i18n:
```

Reply:
Nice, of course docker is always our friend. Thanks @sonalkr132!

Reply:
lol. check this:
```
# select full_name, number, canonical_number from versions where rubygem_id = 159003;
       full_name       |  number   | canonical_number 
-----------------------+-----------+------------------
 algorithmix-0.0.0.6   | 0.0.0.6   | 0.0.0.6
 algorithmix-0.0.0.0   | 0.0.0.0   | .dedup
 algorithmix-0         | 0         | .dedup
 algorithmix-0.0.0.3   | 0.0.0.3   | 0.0.0.3
 algorithmix-0.0.0.4   | 0.0.0.4   | 0.0.0.4
 algorithmix-0.0.0.5   | 0.0.0.5   | 0.0.0.5
 algorithmix-0.0.0.0.0 | 0.0.0.0.0 | .dedup
 algorithmix-0.0.0.4.1 | 0.0.0.4.1 | 0.0.0.4.1
 algorithmix-0.0.0.7   | 0.0.0.7   | 0.0.0.7
 algorithmix-0.0.0.1   | 0.0.0.1   | 0.0.0.1
 algorithmix-00        | 00        | .dedup
 algorithmix-0.0.0.2   | 0.0.0.2   | 0.0.0.2
 algorithmix-0.0.0     | 0.0.0     | 
(13 rows)
```
not covered by our script at the moment.

Reply:
Nice catch!

Regarding the faker 2.0 vs 2.0.0 thing, I was really intrigued about why docker was giving me different results than my local system. It turns out I'm using rubygems & bundler HEAD locally and this seems to have changed since the latest prereleases. I'll do a full bisect tomorrow to figure this out, but the new behaviour seems more consistent since it's the same thing `gem install` does.

Reply:
Ok, I was to curious and don't plan to do any work in the weekend so I bisected it to https://github.com/rubygems/rubygems/commit/0bcf7492d2d633be5b1d71f976f74216b37e08ea. I think [this line in particular](https://github.com/rubygems/rubygems/commit/0bcf7492d2d633be5b1d71f976f74216b37e08ea#diff-aeb9edfaf1badd9cf55c339b60061eeb3d402b2b2182359a21e33e999fc302dcL85).

Anyways, not sure if this changes anything regarding this issue and whether it's worth to keep digging :man_shrugging:.

Reply:
AFAIU, for users who have the exact versions in Gemfile.lock, they will continue to depend on exact versions.

btw, we would have to expunge these versions from our DB besides yanking them, which would be a first for rubygems.org

Reply:
> AFAIU, for users who have the exact versions in Gemfile.lock, they will continue to depend on exact versions.

This is interesting, but it seems like no. In the edge cases of these particular duplications, after upgrading to the latest bundler release, faker 2.0.0 will no longer be installable. Even if you have `2.0.0` already installed (with a previous bundler version) and your lockfile generated to lock to that, running `bundle install` again with the newest bundler will detect faker-2.0 as not installed, and install and use that version.

> btw, we would have to expunge these versions from our DB besides yanking them, which would be a first for rubygems.org

Why is this? Is it because if we yank one of the duplications, all versions with the same canonical number will be interpreted as yanked too for some reason?

Reply:
> running bundle install again with the newest bundler will detect faker-2.0 as not installed, and install and use that version

wow. I hope we have checked that this doesn't have other side effects besides this edge case. It kinda goes against my understanding of the lock file.

> btw, we would have to expunge these versions from our DB besides yanking them, which would be a first for rubygems.org

I just wanted to add this clarification against the initial proposal of yanking versions to make the DB normal. When a version is yanked, we don't delete the version record from the table. we just update one of the columns to mark it unindexed/yanked. This would mean that we will have the duplicate records in the DB even after yank.

Reply:
> wow. I hope we have checked that this doesn't have other side effects besides this edge case. It kinda goes against my understanding of the lock file.

I don't expect it to have other bad consequences, but I could be wrong. This is definitely one I hadn't noticed, hopefully there's no more surprises :pray:.

> I just wanted to add this clarification against the initial proposal of yanking versions to make the DB normal. When a version is yanked, we don't delete the version record from the table. we just update one of the columns to mark it unindexed/yanked. This would mean that we will have the duplicate records in the DB even after yank.

Oh, right, of course, yanking the version wouldn't affect the uniqueness of the `[name, version, platform]` triple, we'd need to add `yanked_at` to the constraint if we want to go that way, and if we yank more than one duplication at the same time, that wouldn't even work...

Reply:
Please let me know if the PR can use any further changes. 
As it stands, it will ensure that all canonical_numbers are unique. check the added rake task.

Reply:
Looks good to me @sonalkr132, thanks for working on this!